src/types/sgrStack.cpp

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "inc/sgrStack.hpp"
+
+using namespace Microsoft::Console::VirtualTerminal::DispatchTypes;
+
+namespace Microsoft::Console::VirtualTerminal
+{
+    SgrStack::SgrStack() noexcept :
+        _nextPushIndex{ 0 },
+        _numSavedAttrs{ 0 }
+    {
+    }
+
+    void SgrStack::Push(const TextAttribute& currentAttributes,
+                        const VTParameters options) noexcept
+    {
+        AttrBitset validParts;
+
+        try
+        {
+            if (options.empty())
+            {
+                // We save all current attributes.
+                validParts.set(static_cast<size_t>(SgrSaveRestoreStackOptions::All));
+            }
+            else
+            {
+                // Each option is encoded as a bit in validParts. All options (that fit) are
+                // encoded; options that aren't supported are ignored when read back (popped).
+                // So if you try to save only unsupported aspects of the current text
+                // attributes, you'll do what is effectively an "empty" push (the subsequent
+                // pop will not change the current attributes), which is the correct behavior.
+
+                for (size_t i = 0; i < options.size(); i++)
+                {
+                    // TRICKY: The VTParameter could be a fake/materialized parameter, in
+                    // which case it has a default value, and the default value of a
+                    // VTParameter depends on what you are assigning to: if you assign /
+                    // convert directly to a size_t, then the default is "1", but for other
+                    // size_t-sized types, the default is "0". We want "default is 0", so we
+                    // convert to a SgrSaveRestoreStackOptions first.
+                    const SgrSaveRestoreStackOptions option = options.at(i);

```

Comment:
You can also use `options.at(i).value_or(0)`, right? Just to be explicit about it :smile: (presuming I've understood correctly)

Reply:
Yeah. It would make more sense to use `value_or(0)` here if that's how it's meant to be interpreted.

Reply:
Much better that way--thanks.